format/ComputeIR.fbs

```diff
@@ -0,0 +1,348 @@
+}
+
+/// An expression is one of
+/// - a Literal datum
+/// - a reference to a Field from a Relation
+/// - a call to a named function
+/// On evaluation, an Expression will have either array or scalar shape.
+union ExpressionImpl {
+  Literal, FieldRef, Call
+}
+
+table Expression {
+  // Ideally we'd simply have `union Expression { Literal, FieldRef, Call }`
+  // but not all generators support vectors of unions so we provide minimal
+  // indirection to support them.
+  impl: ExpressionImpl (required);
+}
+
+union Shape {
+  Array, Scalar
+}
+
+table Scalar {}
+
+table Array {
+  /// Number of slots.
+  length: long;
+}
+
+table Literal {
+  /// Shape of this literal.
+  ///
+  /// Note that this is orthogonal to type and refers to the number
+  /// of rows spanned by this Literal - a Literal may be Scalar shaped
+  /// with multiple "columns" if the type happens to be Struct.
+  shape: Shape (required);
+
+  /// The type of this literal. Field is used instead of Type to pick
+  /// up child fields, dictionary encoding, etc.
+  field: Field (required);
+
+  /// Buffers containing N elements of arrow-formatted data, where N
+  /// is Array.length if shape is Array or 1 if shape is Scalar.
+  /// XXX this can be optimized for trivial scalars later
+  buffers: [InlineBuffer];
+
+  /// If (and only if) this Literal has dictionary type, this field dictionary
+  /// into which the literal's indices refer.
+  dictionary: Literal;
+}
+
+table FieldRef {
+  /// A sequence of field names to allow referencing potentially nested fields
+  path: [string];
+
+  /// For Expressions which might reference fields in multiple Relations,
+  /// this index may be provided to indicate which Relation's fields
+  /// `path` points into. For example in the case of a join,
+  /// 0 refers to the left relation and 1 to the right relation.
+  relation_index: int;
+
+  /// The type of the referenced Field. Field is used instead of Type to pick
+  /// up child fields, dictionary encoding, etc.
+  field: Field;
+}
+
+/// A canonical (probably SQL equivalent) function
+enum CanonicalFunctionId : uint32 {
+  // logical
+  And,
+  Not,
+  Or,
+
+  // arithmetic
+  Add,
+  Subtract,
+  Multiply,
+  Divide,
+  Power,
+  AbsoluteValue,
+  Negate,
+  Sign,
+
+  // comparison
+  Equal,
+  NotEqual,
+  Greater,
+  GreaterOrEqual,
+  Less,
+  LessOrEqual,
+
+  // aggregations
+  All,
+  Any,
+  Count,
+  Mean,
+  Min,
+  Max,
+  Mode,
+  Product,
+  Sum,
+  Tdigest,
+  Quantile,
+  Variance,
+  StandardDeviation,
+}
+
+table CanonicalFunction {
+  id: CanonicalFunctionId;
+}
+
+table NonCanonicalFunction {
+  name_space: string (required);
+  name: string (required);
+}
+
+union Function {
+  CanonicalFunction, NonCanonicalFunction
+}
+
+table Call {
+  /// The function whose invocation this Call represents.
+  function: Function (required);
+
+  /// Parameters for `function_name`; content/format may be unique to each
+  /// value of `function_name`.
+  options: InlineBuffer;
```

Comment:
Per Operation below, I might prefer to have the opaque options only in the NonCanonicalFunction table — in the CanonicalFunction, if needed — may not be yet — we could add a union-of-tables providing function-specific options without the need for InlineBuffer